### ISSUE TITLE
Question about your powerful code

### ISSUE THREAD

**Issue by TangLaoya** [stk-kriging/complex-rational-interpolation#4]
Dear authors,

Thank you very much for providing so powerful matlab to fit the given curve with few sampling points. I tested your code (AdapApprox), and the precision is really the best, however, it is somewhat slow for some complex curves fitting (comparing to other method you provided). In addition, for the data I provided, the code displayed the following error information:

_A is not even close to positive definite_

in 'tune_model'.

Could you please help me to take a look at it?

BTW: do you also have any good idea to implement the adaptive frequency sampling together with your curve fitting? I try the code by Prof. D. Pradovera but also failed.

(see here: 
https://github.com/pradovera/gLoewner/issues/1)

Thanks,
Tang Laoya

[mytest.zip](https://github.com/user-attachments/files/18034590/mytest.zip)

**Comment by jbect** [comment]
Dear Tang,

Your script does not run, because the option structure `opts`  on line 34 is not defined.

Please provide a self-contained script that reproduces the problem you observe.

**Comment by jbect** [comment]
About execution speed:

Yes, our method is indeed slower, in general, than the other methods we are comparing to in the paper.

But please keep in mind that is is essentially a prototype implementation for research: it can certainly be improved as far as the speed of execution is concerned.

In addition, we believe that the main target for this method is the case where the underlying computer model is very expensive to evaluate (e.g., a finite element model), in which case the computational overhead is more acceptable.

Anyway: feel free to optimize the code and contribute back through a PR!

**Comment by jbect** [comment]
> Your script does not run, because the option structure `opts` on line 34 is not defined.

Ok, I have managed to reproduce your error using `opts = init_opts()` to initialize with default options.

I will look into it.

**Comment by jbect** [comment]
> I will look into it.

I think I found it.

Some of your input points (frequencies) are very close to each other.  This results in a covariance matrix that is nearly degenerate (which, in turns, causes `stk_cholcov` to fail).

```
>> xi(1:20,1)

ans =

   1.0e+08 *

   0.000000010000000
   0.000017782794100
   0.018000000000000
   0.070000000000000
   0.155000000000000
   0.370000000000000
   0.630000000000000
   0.890000000000000
   1.150000000000000
   1.410000000000000
   1.670000000000000
   1.930000000000000
   2.190000000000000
   2.450000000000000
   2.710000000000000
   2.970000000000000
   3.230000000000000
   3.490000000000000
   3.750000000000000
   4.010000000000000
```

**Comment by TangLaoya** [comment]
That's great, thank you very much!
Yes, just as you say, currently my model is very complex and the calculated
method is 3-d FEM in frequency domain. However, I have put it to the large
cluster and there are tens of CPU cores and several TB of memory, so I can
launch tens of solvers a time, can your algorithm suitable for this process:

1) generate N frequency points (evenly, usually larger than the number of
solvers M can be launched) and launch M solvers, the left frequency points
are saved at list Freq to be calculated;
2) collect solver results one by one (a S parameter matrix one frequency
point) and when the number reaches to K, launch your code to fit the curve
and calculate the next best frequency point(s), and then put it to the list
Freq;
3) when any solver is finished, pick up a frequency point from list Freq
and launch the solver again;
4) continue to collect solver results to update the 'measured results', and
launch your code to update the curve, and get the next best frequency
point(s), and put it to the list Freq;
5) the iteration converges when the maximal error is smaller than the given
threshold.


Thanks,
Tang Laoya

**Comment by TangLaoya** [comment]
> > I will look into it.
> 
> I think I found it.
> 
> Some of your input points (frequencies) are very close to each other. This results in a covariance matrix that is nearly degenerate (which, in turns, causes `stk_cholcov` to fail).
> 
> ```
> >> xi(1:20,1)
> 
> ans =
> 
>    1.0e+08 *
> 
>    0.000000010000000
>    0.000017782794100
>    0.018000000000000
>    0.070000000000000
>    0.155000000000000
>    0.370000000000000
>    0.630000000000000
>    0.890000000000000
>    1.150000000000000
>    1.410000000000000
>    1.670000000000000
>    1.930000000000000
>    2.190000000000000
>    2.450000000000000
>    2.710000000000000
>    2.970000000000000
>    3.230000000000000
>    3.490000000000000
>    3.750000000000000
>    4.010000000000000
> ```

Yes, it is the original full data I got after tens of days calculation, my target is to select smallest number of frequency points to fit the original curve with target error threshold (should allow tens of solvers be launched at the same time).
